src/Geocoder/Provider/AbstractHttpProvider.php

```diff
@@ -18,27 +21,28 @@
 class AbstractHttpProvider extends AbstractProvider
 {
     /**
-     * @var HttpAdapterInterface
+     * @var HttpClient
      */
-    private $adapter;
+    private $client;
 
     /**
-     * @param HttpAdapterInterface $adapter An HTTP adapter
+     * @param HttpClient $client An HTTP adapter
      */
-    public function __construct(HttpAdapterInterface $adapter)
+    public function __construct(HttpClient $client = null, MessageFactory $factory = null)
     {
         parent::__construct();
 
-        $this->adapter = $adapter;
+        $this->client = $client ?: HttpClientDiscovery::find();
+        $this->factory = $factory ?: MessageFactoryDiscovery::find();
```

Comment:
One of the way we have done this for not having discovery, it to put discovery in suggest package then on the constructor we check the existence of the DiscoveryClass, if it has not been found we throw a \LogicException with a specific error which will be in our documentation for easy googling on that, here is an example : 

```php
if (null === $responseFactory && !class_exists('\Http\Discovery\MessageFactoryDiscovery')) {
    throw new \LogicException('No response factory provided and no discovery service is present to guess it, maybe you need to install php-http/discovery package?');
}
```